### PR TITLE
bcd: add interface for passing arguments

### PR DESCRIPTION
**Pull request by dhobsd**
Argument order is preserved to be the order of the calls to bcd_arg.
The utility of this is to be able to specify arbitrary arguments to the
tracing tool, as various arguments have utility outside of kv-pairs and
specifying a thread to trace.

### CODE REVIEW

File: src/bcd.c [hunk 1]
```diff
@@ -87,6 +90,16 @@ struct bcd_kv {
 static LIST_HEAD(, bcd_kv) bcd_kv_list = LIST_HEAD_INITIALIZER(&bcd_kv_list);
 static size_t bcd_kv_length;
 static size_t bcd_kv_count;
+
+struct bcd_arg {
+	LIST_ENTRY(bcd_arg) linkage;
+	const char *arg;
+};
+static LIST_HEAD(, bcd_arg) bcd_arg_list = LIST_HEAD_INITIALIZER(&bcd_arg_list);
```
Comment (abelmathew) [inline src/bcd.c]: Why not TAILQ? Removes the need for `arg_tail` as well.

File: src/bcd.c [hunk 2]
```diff
@@ -581,6 +602,91 @@ bcd_kv_set(struct bcd_session *session, struct bcd_packet *packet)
...
+static ssize_t
+bcd_arg_get(char **output, size_t n_output, bcd_error_t *error)
+{
+	struct bcd_arg *cursor;
+	size_t limit = n_output;
+	size_t i = 0;
+
+	if (bcd_arg_count == 0) {
+		return 0;
+	}
```
Comment (abelmathew) [inline src/bcd.c]: Style: the rest of the code leaves out curly braces for conditionals if the code contained is one line.

File: src/bcd.c [hunk 3]
```diff
@@ -581,6 +602,91 @@ bcd_kv_set(struct bcd_session *session, struct bcd_packet *packet)
...
+		return bcd_error(BCD_EVENT_METADATA, session,
+		    "internal memory allocation error");
+	}
+
+	if (bcd_arg_count == 0) {
+		TAILQ_INIT(&bcd_arg_list);
+	} else {
+		struct bcd_arg *tmp;
+
+		TAILQ_FOREACH_SAFE(cursor, &bcd_arg_list, linkage, tmp) {
```
Comment (abelmathew) [inline src/bcd.c]: Why do you need `FOREACH_SAFE` here? You break after cleanup so as far as I can tell, TAILQ_FOREACH should be sufficient.

File: include/bcd.h [hunk 4]
```diff
@@ -238,11 +238,17 @@ const char *bcd_error_message(const bcd_error_t *);
...
- * Make key-value association.
+ * Make key-value association. KV-pairs are LIFO-ordered.
  */
 int bcd_kv(bcd_t *, const char *, const char *, bcd_error_t *);
 
 /*
+ * Add an argument to the command line. Arguments are presented to the tracing
+ * process in order.
+ */
+int bcd_arg(bcd_t *, const char *, bcd_error_t *);
```
Comment (abelmathew) [inline include/bcd.h]: I forgot to ask this before, why do we want FIFO ordering for arguments?

File: src/bcd.c (hunk 3 above)
Comment (dhobsd) [inline src/bcd.c]: Overthinking it.

File: include/bcd.h (hunk 4 above)
Comment (dhobsd) [inline include/bcd.h]: Argument order matters when calling into e.g. `coroner backtrace` and it feels unnatural to push arguments onto a stack that is applied in reverse order -- especially when the arguments are part of some configuration tunable. But either way, I think POLA applies. I don't see any reason to make argument order non-deterministic.

Comment (abelmathew) [inline include/bcd.h]: ah, didn't think about `coroner backtrace`

**Pull request merged**